### PR TITLE
fix(animations): use a lightweight renderer for non-animation components

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

File: packages/animations/browser/src/render/transition_animation_engine.ts [hunk 1]
```diff
@@ -13,12 +13,13 @@ import {AnimationTransitionInstruction} from '../dsl/animation_transition_instru
...
 import {ElementInstructionMap} from '../dsl/element_instruction_map';
 import {AnimationStyleNormalizer} from '../dsl/style_normalization/animation_style_normalizer';
-import {ENTER_CLASSNAME, LEAVE_CLASSNAME, LEAVE_SELECTOR, NG_ANIMATING_CLASSNAME, NG_TRIGGER_CLASSNAME, NG_TRIGGER_SELECTOR, copyObj, eraseStyles, iteratorToArray, setStyles} from '../util';
+import {ENTER_CLASSNAME, LEAVE_CLASSNAME, LEAVE_SELECTOR, NG_ANIMATING_CLASSNAME, NG_TRIGGER_CLASSNAME, NG_TRIGGER_SELECTOR, copyObj, eraseStyles, setStyles} from '../util';
 
 import {AnimationDriver} from './animation_driver';
 import {getOrSetAsInMap, listenOnPlayer, makeAnimationEvent, normalizeKeyframes, optimizeGroupPlayer} from './shared';
 
 const EMPTY_PLAYER_ARRAY: AnimationPlayer[] = [];
+const ANIMATE_EPOCH_ATTR = 'ng-animate-epoch';
```
Comment (tbosch) [inline packages/animations/browser/src/render/transition_animation_engine.ts]: maybe a better name?

File: packages/platform-browser/animations/src/animation_renderer.ts [hunk 2]
```diff
@@ -29,18 +30,29 @@ export class AnimationRendererFactory implements RendererFactory2 {
   }
 
   createRenderer(hostElement: any, type: RendererType2): Renderer2 {
-    let delegate = this.delegate.createRenderer(hostElement, type);
-    if (!hostElement || !type || !type.data || !type.data['animation']) return delegate;
+    // cache the delegates to find out which cached delegate can
+    // be used by which cached renderer
+    const delegate = this.delegate.createRenderer(hostElement, type);
+    let renderer: BaseAnimationRenderer|undefined = this._rendererCache.get(delegate);
```
Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: move `renderer` inside of the `if`, as it is not needed anywhere else.

File: packages/platform-browser/animations/src/animation_renderer.ts [hunk 3]
```diff
@@ -79,89 +90,113 @@ export class AnimationRendererFactory implements RendererFactory2 {
...
 
-  whenRenderingDone(): Promise<any> { return this._engine.whenRenderingDone(); }
+  whenRenderingDone(): Promise<any> { return this.engine.whenRenderingDone(); }
 }
 
-export class AnimationRenderer implements Renderer2 {
-  public destroyNode: ((node: any) => any)|null = null;
-  public microtaskCount: number = 0;
+export class BaseAnimationRenderer implements Renderer2 {
+  public namespaceId: string = '';
```
Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: take the `namespaceId` in the constructor, and make it `protected`.

File: packages/platform-browser/animations/src/animation_renderer.ts [hunk 4]
```diff
@@ -79,89 +90,113 @@ export class AnimationRendererFactory implements RendererFactory2 {
...
     this.delegate.setStyle(el, style, value, flags);
   }
 
-  removeStyle(el: any, style: string, flags: RendererStyleFlags2): void {
+  removeStyle(el: any, style: string, flags?: RendererStyleFlags2|undefined): void {
     this.delegate.removeStyle(el, style, flags);
   }
 
+  setProperty(el: any, name: string, value: any): void {
+    if (name.charAt(0) != '@') {
```
Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: I think this check is not needed...

File: packages/platform-browser/animations/src/animation_renderer.ts [hunk 5]
```diff
@@ -79,89 +90,113 @@ export class AnimationRendererFactory implements RendererFactory2 {
...
+    }
+  }
+
   setValue(node: any, value: string): void { this.delegate.setValue(node, value); }
 
-  appendChild(parent: any, newChild: any): void {
-    this.delegate.appendChild(parent, newChild);
-    this._engine.onInsert(this._namespaceId, newChild, parent, false);
+  listen(target: any, eventName: string, callback: (event: any) => boolean | void): () => void {
+    if (eventName.charAt(0) == '@') {
```
Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: Why is this logic needed for the base renderer?

Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: I.e. it probably should go into the regular `AnimationRenderer`.

File: packages/platform-browser/animations/src/animation_renderer.ts [hunk 6]
```diff
@@ -12,12 +12,13 @@ import {Injectable, NgZone, Renderer2, RendererFactory2, RendererStyleFlags2, Re
...
-  private _currentFlushId: number = 1;
+  private _microtaskId: number = 1;
   private _animationCallbacksBuffer: [(e: any) => any, any][] = [];
+  private _rendererCache = new Map<Renderer2, BaseAnimationRenderer>();
 
   constructor(
-      private delegate: RendererFactory2, private _engine: AnimationEngine, private _zone: NgZone) {
-    _engine.onRemovalComplete = (element: any, delegate: any) => {
+      private delegate: RendererFactory2, private engine: AnimationEngine, private _zone: NgZone) {
+    engine.onRemovalComplete = (element: any, delegate: any) => {
```
Comment (tbosch) [inline packages/platform-browser/animations/src/animation_renderer.ts]: type the `delegate` to be a `RendererV2`.

**Review by tbosch — APPROVED** [review]
Reviewed the changes to core and the renderer.

**Review by mhevery — APPROVED** [review]

**Pull request merged**